modules/cdp-engine/src/lib.rs

```diff
@@ -1334,47 +1329,127 @@ impl<T: Config> LiquidateCollateral<T::AccountId> for LiquidateViaDex<T> {
 			)?;
 		}
 
-		Ok(())
+		let swap_price = Price::saturating_from_rational(actual_target_amount, actual_supply_collateral);
+		let ratio = swap_price
+			.checked_div(&collateral_oracle_price)
+			.ok_or(ArithmeticError::DivisionByZero)?;
+		Ok(ratio)
+	}
+
+	fn liquidate_via_contract(
+		contract: EvmAddress,
+		currency_id: CurrencyId,
+		amount: Balance,
+		target_stable_amount: Balance,
+		collateral_oracle_price: Price,
+	) -> Result<Ratio, DispatchError> {
+		let collateral = currency_id
+			.erc20_address()
+			.ok_or(Error::<T>::CollateralContractNotFound)?;
+		let repay_dest = Self::evm_address();
+		let repay_dest_account_id = Self::account_id();
+		let stable_coin = T::GetStableCurrencyId::get();
+		let repay_dest_balance = CurrencyOf::<T>::free_balance(stable_coin, &repay_dest_account_id);
+
+		T::LiquidationEvmBridge::liquidate(
+			InvokeContext {
+				contract,
+				sender: repay_dest,
+				origin: contract,
+			},
+			collateral,
+			repay_dest,
+			amount,
+			target_stable_amount,
+		)?;
+
+		let repayment =
+			CurrencyOf::<T>::free_balance(stable_coin, &repay_dest_account_id).saturating_sub(repay_dest_balance);
+		let contract_account_id = T::EvmAddressMapping::get_account_id(&contract);
+		if repayment >= target_stable_amount {
+			// sufficient repayment, transfer collateral to contract and notify
+			<T as Config>::CDPTreasury::withdraw_collateral(&contract_account_id, currency_id, amount)?;
+			// notify liquidation success
+			T::LiquidationEvmBridge::on_collateral_transfer(
+				InvokeContext {
+					contract,
+					sender: repay_dest,
+					origin: contract,
+				},
+				collateral,
+				target_stable_amount,
+			);
+			let liquidated_price = Price::saturating_from_rational(repayment, amount);
+			let ratio = liquidated_price
+				.checked_div(&collateral_oracle_price)
+				.ok_or(ArithmeticError::DivisionByZero)?;
+			return Ok(ratio);
+		} else if repayment > 0 {
+			// insufficient repayment, refund
+			CurrencyOf::<T>::transfer(stable_coin, &repay_dest_account_id, &contract_account_id, repayment)?;
```

Comment:
I think we can just revert the previous liquidate call using `with_tranasction` so no need to refund